pkg/util/inplaceupdate/inplace_update_defaults.go

```diff
@@ -149,18 +141,27 @@ func defaultCalculateInPlaceUpdateSpec(oldRevision, newRevision *apps.Controller
 // If the imageID in containerStatuses has not been changed, we assume that kubelet has not updated
 // containers in Pod.
 func DefaultCheckInPlaceUpdateCompleted(pod *v1.Pod) error {
+	if _, isInGraceState := appspub.GetInPlaceUpdateGrace(pod); isInGraceState {
+		return fmt.Errorf("still in grace period of in-place update")
+	}
+
 	inPlaceUpdateState := appspub.InPlaceUpdateState{}
 	if stateStr, ok := appspub.GetInPlaceUpdateState(pod); !ok {
 		return nil
 	} else if err := json.Unmarshal([]byte(stateStr), &inPlaceUpdateState); err != nil {
 		return err
 	}
 
-	// this should not happen, unless someone modified pod revision label
-	//if inPlaceUpdateState.Revision != pod.Labels[apps.StatefulSetRevisionLabel] {
-	//	return fmt.Errorf("currently revision %s not equal to in-place update revision %s",
-	//		pod.Labels[apps.StatefulSetRevisionLabel], inPlaceUpdateState.Revision)
-	//}
+	runtimeContainerMetaSet, err := appspub.GetRuntimeContainerMetaSet(pod)
+	if err != nil {
+		return err
+	}
+	if runtimeContainerMetaSet != nil {
+		if checkAllContainersHashConsistent(pod, runtimeContainerMetaSet) {
+			klog.V(5).Infof("Check Pod %s/%s in-place update completed for all container hash consistent", pod.Namespace, pod.Name)
+			return nil
+		}
```

Comment:
Shouldn't we just return error here?

Reply:
Sometimes users might not have kruise-daemon on each node, e.g. some nodes are created by VK. So we can't totally rely on the annotation report.

Reply:
If the kruise-daemon used to report meta annotation and now it has broken for some reason, we expect it as no effect on pods update as possible.